FEXCore/Source/Interface/Core/Frontend.cpp

```diff
@@ -64,33 +67,75 @@ static uint32_t MapVEXToReg(uint8_t vvvv, bool HasXMM) {
   }
 }
 
-Decoder::Decoder(FEXCore::Context::ContextImpl* ctx)
-  : CTX {ctx}
-  , OSABI {ctx->SyscallHandler ? ctx->SyscallHandler->GetOSABI() : FEXCore::HLE::SyscallOSABI::OS_UNKNOWN}
-  , PoolObject {ctx->FrontendAllocator, sizeof(FEXCore::X86Tables::DecodedInst) * DefaultDecodedBufferSize} {}
+Decoder::Decoder(FEXCore::Core::InternalThreadState* Thread)
+  : Thread {Thread}
+  , CTX {static_cast<FEXCore::Context::ContextImpl*>(Thread->CTX)}
+  , OSABI {CTX->SyscallHandler ? CTX->SyscallHandler->GetOSABI() : FEXCore::HLE::SyscallOSABI::OS_UNKNOWN}
+  , PoolObject {CTX->FrontendAllocator, sizeof(FEXCore::X86Tables::DecodedInst) * DefaultDecodedBufferSize} {}
 
 Decoder::~Decoder() {
   PoolObject.UnclaimBuffer();
 }
 
+bool Decoder::CheckRangeExecutable(uint64_t Address, uint64_t Size) {
+  // Treat FEX-internal X86 callbacks as always executable
+  if (EntryPoint == CTX->X86CodeGen.CallbackReturn) {
+    return true;
+  }
+
+  while (Address < ExecutableRangeBase || Address + Size > ExecutableRangeEnd) {
+    auto RangeInfo = CTX->SyscallHandler->QueryGuestExecutableRange(Thread, Address);
+    ExecutableRangeBase = Address;
+    ExecutableRangeEnd = Address + RangeInfo.Size;
+
+    if (RangeInfo.Size == 0) {
+      return false;
+    }
+
+    if (Size > RangeInfo.Size) {
+      Size -= RangeInfo.Size;
+      Address += RangeInfo.Size;
+    }
+  }
+
+  return true;
+}
+
 uint8_t Decoder::ReadByte() {
-  uint8_t Byte = InstStream[InstructionSize];
   LOGMAN_THROW_A_FMT(InstructionSize < MAX_INST_SIZE, "Max instruction size exceeded!");
+  uint8_t Byte = PeekByte(0);
   Instruction[InstructionSize] = Byte;
   InstructionSize++;
   return Byte;
 }
 
-uint8_t Decoder::PeekByte(uint8_t Offset) const {
-  uint8_t Byte = InstStream[InstructionSize + Offset];
+uint8_t Decoder::PeekByte(uint8_t Offset) {
+  uint8_t Byte = 0;
+  uint64_t ByteAddress = reinterpret_cast<uint64_t>(InstStream + InstructionSize + Offset);
+  if (CheckRangeExecutable(ByteAddress, 1)) {
+    Byte = InstStream[InstructionSize + Offset];
+  } else {
+    HitNonExecutableRange = true;
+    // Pretend we read 0, the main decode loop will see HitNonExecutableRange and rollback the instruction.
```

Comment:
In my code caching branch (which integrates this PR), I keep running into different issues that lead me back to this line. The bugs are on the code caching branch, but they are difficult to debug because returning 0 here silently swallows potential errors and defers them until an unspecific later point in time. For example:

* I had implemented QueryExecutableRange implemented slightly incorrectly, causing some instructions to be replaced with null bytes
* I compiled valid x86 code that simply hadn't be marked as `PROT_EXEC` yet, causing FEX to just emit a panic block

Are there any safeguards we can put in place around this code to detect invalid scenarios more eagerly?

Reply:
> warning in the log

I'm not too interested in adding likely false-positives to the log either, but more in situations where we definitely know something went wrong and hence can loudly fail. This could well be a heuristic approach, i.e. any combination of trivial `if` statements that doesn't hinder normal multiblock operation but eases debugging.

> I think your issue itself is effectively desired behaviour if multiblock is enabled

Is this unconditionally true? What about the case where we haven't successfully decoded any instructions yet - is there a valid use case for allowing this?

> Perhaps if multiblock is disabled and the block hits a decode failure we could print?

Is there a reason not to panic in this case?